packages/@aws-cdk/tmp-toolkit-helpers/src/api/refactoring/index.ts

```diff
@@ -0,0 +1,415 @@
+      after.filter((a) => !common.some((c) => equalLocations(a, c))),
+    ];
+  }
+
+  return result;
+}
+
+function equalLocations(a: ResourceLocation, b: ResourceLocation): boolean {
+  return a.logicalResourceId === b.logicalResourceId && a.stack.stackName === b.stack.stackName;
+}
+
+/**
+ * For each hash, identifying a single resource, zip the two lists of locations,
+ * to produce a pair of arrays of locations, in the same order as the parameters.
+ */
+function zip(
+  m1: Record<string, ResourceLocation[]>,
+  m2: Record<string, ResourceLocation[]>,
+): Record<string, [ResourceLocation[], ResourceLocation[]]> {
+  const result: Record<string, [ResourceLocation[], ResourceLocation[]]> = {};
+
+  for (const [hash, locations] of Object.entries(m1)) {
+    if (hash in m2) {
+      result[hash] = [locations, m2[hash]];
+    } else {
+      result[hash] = [locations, []];
+    }
+  }
+
+  for (const [hash, locations] of Object.entries(m2)) {
+    if (!(hash in m1)) {
+      result[hash] = [[], locations];
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Computes a list of pairs [digest, location] for each resource in the stack.
+ */
+function index(stack: BasicStack): [string, ResourceLocation][] {
+  const digests = computeResourceDigests(stack.template);
+
+  return Object.entries(digests).map(([logicalId, digest]) => {
+    const location: ResourceLocation = new ResourceLocation(stack, logicalId);
+    return [digest, location];
+  });
+}
+
+/**
+ * Detects refactor mappings by comparing the stacks in the given environment
+ * with the stacks deployed to the same environment.
+ *
+ * @param stacks - The stacks to compare.
+ * @param sdkProvider - The SDK provider to use for fetching deployed stacks.
+ * @returns A promise that resolves to an array of resource mappings.
+ */
+export async function detectRefactorMappings(
+  stacks: BasicStack[],
+  sdkProvider: SdkProvider,
+): Promise<ResourceMapping[]> {
+  const stackGroups: Map<string, [BasicStack[], BasicStack[]]> = new Map();
+
+  // Group stacks by environment
+  for (const stack of stacks) {
+    const environment = stack.environment;
+    const key = hashObject(environment);
+    if (stackGroups.has(key)) {
+      stackGroups.get(key)![1].push(stack);
+    } else {
+      // The first time we see an environment, we need to fetch all stacks deployed to it.
+      const before = await getDeployedStacks(sdkProvider, environment);
+      stackGroups.set(key, [before, [stack]]);
+    }
+  }
+
+  const result: ResourceMapping[] = [];
+  for (const [_, [before, after]] of stackGroups) {
+    result.push(...computeMappings(before, after));
+  }
+  return result;
+}
+
+async function getDeployedStacks(sdkProvider: SdkProvider, environment: cxapi.Environment): Promise<BasicStack[]> {
+  const cfn = (await sdkProvider.forEnvironment(environment, Mode.ForReading)).sdk.cloudFormation();
+  const summaries: StackSummary[] = [];
+  await paginateSdkCall(async (nextToken) => {
```

Comment:
Doesn't the SDK abstraction we return paginate already? Did we or did we not do that?